File: agent-runtime/location_intelligence_contracts.py
"""Typed contracts for Location Intelligence research and synthesis."""

from __future__ import annotations

from decimal import Decimal
from typing import Annotated, Literal
from uuid import UUID

from pydantic import Field, model_validator

from contracts import AgentInvocationEnvelope, ContractModel, StableCode
from reference_evidence import ReferenceObservation
from supplied_brief_contracts import BriefConflict

RESEARCH_OPERATION = "LOCATION_RESEARCH_PLAN"
SYNTHESIS_OPERATION = "LOCATION_SYNTHESIS"


class LocationAudienceSegment(ContractModel):
    segment_id: UUID
    name: Annotated[str, Field(min_length=1, max_length=300)]
    description: Annotated[str, Field(min_length=1, max_length=2_000)]
    geographies: tuple[Annotated[str, Field(min_length=1, max_length=300)], ...]
    classification: StableCode
    evidence_item_ids: tuple[UUID, ...] = ()
    reference_observation_ids: tuple[UUID, ...] = ()
    is_target: bool


class LocationPoiCategoryOption(ContractModel):
    code: StableCode
    label: Annotated[str, Field(min_length=1, max_length=200)]


class LocationCommercialContext(ContractModel):
    brief_version_id: UUID
    audience_artifact_id: UUID
    audience_artifact_version: int
    client_name: Annotated[str, Field(min_length=1, max_length=500)]
    business_problem: Annotated[str, Field(min_length=1, max_length=4_000)]
    objective: Annotated[str, Field(min_length=1, max_length=4_000)]
    geographies: Annotated[tuple[str, ...], Field(min_length=1, max_length=50)]
    media_requirements: tuple[Annotated[str, Field(min_length=1, max_length=1_000)], ...] = ()
    constraints: tuple[Annotated[str, Field(min_length=1, max_length=1_000)], ...] = ()
    conflicts: tuple[BriefConflict, ...] = ()
    target_segments: tuple[LocationAudienceSegment, ...] = ()
    reference_evidence: tuple[ReferenceObservation, ...] = ()
    available_poi_categories: Annotated[tuple[LocationPoiCategoryOption, ...], Field(min_length=1)]


class LocationResearchPlanRequest(ContractModel):
    operation: Literal["LOCATION_RESEARCH_PLAN"]
    invocation: AgentInvocationEnvelope
    location: LocationCommercialContext

    @model_validator(mode="after")
    def validate_resources(self) -> LocationResearchPlanRequest:
        _require_reference(self.invocation, "BriefVersion", self.location.brief_version_id)
        _require_reference(
            self.invocation,
            "IntelligenceArtifact",
            self.location.audience_artifact_id,
            self.location.audience_artifact_version,
        )
        return self


class PlaceResearchQuery(ContractModel):
    poi_category: StableCode
    anchor_geography: Annotated[str, Field(min_length=2, max_length=300)]
    purpose: Annotated[str, Field(min_length=1, max_length=1_000)]
    priority: Literal["REQUIRED", "PREFERRED"]
    classification: Literal["INFERENCE", "HYPOTHESIS"]
    reference_observation_ids: tuple[UUID, ...] = ()


class LocationResearchPlanArtifact(ContractModel):
    queries: Annotated[tuple[PlaceResearchQuery, ...], Field(max_length=8)]
    rationale: Annotated[str, Field(min_length=1, max_length=4_000)]
    evidence_gaps: tuple[Annotated[str, Field(min_length=1, max_length=1_000)], ...] = ()


class ResolvedPlace(ContractModel):
    place_id: Annotated[str, Field(min_length=1, max_length=300)]
    query: Annotated[str, Field(min_length=2, max_length=200)]
    purpose: Annotated[str, Field(min_length=1, max_length=1_000)]
    name: Annotated[str, Field(min_length=1, max_length=500)]
    address: Annotated[str, Field(min_length=1, max_length=1_000)]
    latitude: Annotated[Decimal, Field(ge=-90, le=90)]
    longitude: Annotated[Decimal, Field(ge=-180, le=180)]
    source_locator: Annotated[str, Field(min_length=1, max_length=2_000)]
    attribution: Annotated[str, Field(min_length=1, max_length=500)]
    geometry_basis: Annotated[str, Field(min_length=1, max_length=100)]


class LocationSynthesisContext(LocationCommercialContext):
    research_plan: LocationResearchPlanArtifact
    resolved_places: tuple[ResolvedPlace, ...] = ()


class LocationSynthesisRequest(ContractModel):
    operation: Literal["LOCATION_SYNTHESIS"]
    invocation: AgentInvocationEnvelope
    location: LocationSynthesisContext

    @model_validator(mode="after")
    def validate_resources(self) -> LocationSynthesisRequest:
        _require_reference(self.invocation, "BriefVersion", self.location.brief_version_id)
        _require_reference(
            self.invocation,
            "IntelligenceArtifact",
            self.location.audience_artifact_id,
            self.location.audience_artifact_version,
        )
        return self


class LocationOpportunityArea(ContractModel):
    name: Annotated[str, Field(min_length=1, max_length=500)]
    geography: Annotated[str, Field(min_length=1, max_length=500)]
    rationale: Annotated[str, Field(min_length=1, max_length=2_000)]
    classification: Literal["INFERENCE", "HYPOTHESIS"]
    place_ids: tuple[Annotated[str, Field(min_length=1, max_length=300)], ...] = ()
    reference_observation_ids: tuple[UUID, ...] = ()
    confidence: Annotated[Decimal, Field(ge=0, le=1)] | None = None
    evidence_gaps: tuple[Annotated[str, Field(min_length=1, max_length=1_000)], ...] = ()


class LocationSynthesisProposal(ContractModel):
    summary: Annotated[str, Field(min_length=1, max_length=4_000)]
    opportunities: Annotated[tuple[LocationOpportunityArea, ...], Field(max_length=20)]
    evidence_gaps: tuple[Annotated[str, Field(min_length=1, max_length=1_000)], ...] = ()


class LocationIntelligenceArtifact(LocationSynthesisProposal):
    research_queries: tuple[PlaceResearchQuery, ...]
    resolved_places: tuple[ResolvedPlace, ...]


def _require_reference(
    invocation: AgentInvocationEnvelope,
    resource_type: str,
    resource_id: UUID,
    version: int | None = None,
) -> None:
    if not any(
        item.resource_type == resource_type
        and item.resource_id == resource_id
        and (version is None or item.version == version)
        for item in invocation.resource_refs
    ):
        raise ValueError(f"The exact {resource_type} resource reference is required.")

File: api/src/Advertified.Commercial.Application/LocationIntelligence/LocationIntelligenceContracts.cs
using Advertified.Commercial.Application.Intelligence;

namespace Advertified.Commercial.Application.LocationIntelligence;

public sealed record LocationAudienceSegmentInput(
    Guid SegmentId,
    string Name,
    string Description,
    IReadOnlyList<string> Geographies,
    string Classification,
    IReadOnlyList<Guid> EvidenceItemIds,
    IReadOnlyList<Guid> ReferenceObservationIds,
    bool IsTarget);

public sealed record LocationPoiCategoryOption(
    string Code,
    string Label);

public sealed record LocationIntelligenceInput(
    CommercialProblemInput Problem,
    Guid AudienceArtifactId,
    long AudienceArtifactVersion,
    IReadOnlyList<LocationAudienceSegmentInput> TargetSegments,
    IReadOnlyList<ReferenceObservationFact> ReferenceEvidence,
    IReadOnlyList<LocationPoiCategoryOption> AvailablePoiCategories);

public sealed record PlaceResearchQueryProposal(
    string PoiCategory,
    string AnchorGeography,
    string Purpose,
    string Priority,
    string Classification,
    IReadOnlyList<Guid> ReferenceObservationIds);

public sealed record LocationResearchPlanProposal(
    IReadOnlyList<PlaceResearchQueryProposal> Queries,
    string Rationale,
    IReadOnlyList<string> EvidenceGaps,
    IReadOnlyList<string> Unknowns,
    IntelligenceInvocationUsage Usage);

public sealed record ResolvedLocationPlace(
    string PlaceId,
    string Query,
    string Purpose,
    string Name,
    string Address,
    decimal Latitude,
    decimal Longitude,
    string SourceLocator,
    string Attribution,
    string GeometryBasis);

public sealed record LocationOpportunityAreaProposal(
    string Name,
    string Geography,
    string Rationale,
    string Classification,
    IReadOnlyList<string> PlaceIds,
    IReadOnlyList<Guid> ReferenceObservationIds,
    decimal? Confidence,
    IReadOnlyList<string> EvidenceGaps);

public sealed record LocationResearchQueryArtifact(
    string PoiCategory,
    string AnchorGeography,
    string Purpose,
    string Priority,
    string Classification,
    IReadOnlyList<Guid> ReferenceObservationIds);

public sealed record ResolvedLocationPlaceArtifact(
    string PlaceId,
    string Query,
    string Purpose,
    string Name,
    string Address,
    decimal Latitude,
    decimal Longitude,
    string SourceLocator,
    string Attribution,
    string GeometryBasis);

public sealed record LocationOpportunityAreaArtifact(
    string Name,
    string Geography,
    string Rationale,
    string Classification,
    IReadOnlyList<string> PlaceIds,
    IReadOnlyList<Guid> ReferenceObservationIds,
    decimal? Confidence,
    IReadOnlyList<string> EvidenceGaps);

public sealed record LocationIntelligenceArtifact(
    string Summary,
    IReadOnlyList<LocationOpportunityAreaArtifact> Opportunities,
    IReadOnlyList<LocationResearchQueryArtifact> ResearchQueries,
    IReadOnlyList<ResolvedLocationPlaceArtifact> ResolvedPlaces,
    IReadOnlyList<string> EvidenceGaps);

public sealed record LocationIntelligenceProposal(
    string Summary,
    IReadOnlyList<LocationOpportunityAreaProposal> Opportunities,
    IReadOnlyList<PlaceResearchQueryProposal> ResearchQueries,
    IReadOnlyList<ResolvedLocationPlace> ResolvedPlaces,
    IReadOnlyList<string> EvidenceGaps,
    IReadOnlyList<string> Unknowns,
    IReadOnlyList<string> Assumptions,
    string Rationale,
    IntelligenceInvocationUsage Usage);

public interface ILocationIntelligenceAgentClient
{
    Task<LocationResearchPlanProposal> PlanResearchAsync(
        LocationIntelligenceInput input,
        CancellationToken cancellationToken);

    Task<LocationIntelligenceProposal> SynthesizeAsync(
        LocationIntelligenceInput input,
        LocationResearchPlanProposal plan,
        IReadOnlyList<ResolvedLocationPlace> resolvedPlaces,
        CancellationToken cancellationToken);
}

public interface ILocationIntelligenceService
{
    Task<IntelligenceArtifactView> AnalyseBriefAsync(
        Guid actorId,
        Guid tenantId,
        Guid briefVersionId,
        CancellationToken cancellationToken);

    Task<IntelligenceArtifactView?> GetLatestAsync(
        Guid actorId,
        Guid tenantId,
        Guid briefVersionId,
        CancellationToken cancellationToken);
}
